File: utils/color_manager.py
"""
Color Management Utility for Energy Platform
Centralized color management system for consistent theming across the application
"""

import json
import os
import logging
from typing import Dict, List, Optional, Union
from pathlib import Path

logger = logging.getLogger(__name__)

class ColorManager:
    """
    Centralized color management system for the entire application
    Supports sectors, models, carriers, and custom color schemes
    """
    
    def __init__(self, app=None):
        self.app = app
        self.config_file = None
        self.colors = {}
        self.default_colors = self._get_default_colors()
        
        if app:
            self.init_app(app)
    
    def init_app(self, app):
        """Initialize with Flask app"""
        self.app = app
        static_folder = app.static_folder or 'static'
        self.config_file = os.path.join(static_folder, 'config', 'colors.json')
        
        # Ensure config directory exists
        os.makedirs(os.path.dirname(self.config_file), exist_ok=True)
        
        # Load existing colors or create default
        self.load_colors()
        
        # Register template functions
        app.jinja_env.globals['get_color'] = self.get_color
        app.jinja_env.globals['get_color_palette'] = self.get_color_palette
        app.jinja_env.globals['get_all_colors'] = self.get_all_colors
    
    def _get_default_colors(self) -> Dict:
        """Define comprehensive default color schemes with enhanced visual appeal"""
        return {
            "sectors": {
                "residential": "#2563EB",      # Vibrant Blue
                "commercial": "#059669",       # Emerald Green
                "industrial": "#DC2626",       # Strong Red
                "transportation": "#7C3AED",   # Rich Purple
                "agriculture": "#16A34A",      # Forest Green
                "public": "#EA580C",           # Warm Orange
                "mining": "#92400E",           # Earth Brown
                "construction": "#0891B2",     # Steel Blue
                "services": "#BE185D",         # Magenta Pink
                "healthcare": "#0D9488",       # Medical Teal
                "education": "#7C2D12",        # Academic Brown
                "hospitality": "#B91C1C",      # Hospitality Red
                "retail": "#1D4ED8",           # Commerce Blue
                "finance": "#6B21A8",          # Finance Purple
                "technology": "#0F766E",       # Tech Teal
                "manufacturing": "#A16207",    # Industrial Amber
                "utilities": "#DC2626",        # Utility Red
                "government": "#1F2937",       # Official Gray
                "military": "#374151",         # Military Gray
                "other": "#6B7280"             # Neutral Gray
            },
            "models": {
                "MLR": "#2563EB",              # Primary Blue
                "SLR": "#059669",              # Success Green
                "WAM": "#F59E0B",              # Warning Amber
                "TimeSeries": "#8B5CF6",       # Time Violet
                "ARIMA": "#EF4444",            # Alert Red
                "Linear": "#06B6D4",           # Linear Cyan
                "Polynomial": "#84CC16",       # Growth Lime
                "Exponential": "#F97316",      # Energy Orange
                "Logarithmic": "#EC4899",      # Analysis Pink
                "Power": "#14B8A6",            # Power Teal
                "Neural": "#A855F7",           # AI Purple
                "RandomForest": "#22C55E",     # Forest Green
                "SVM": "#F43F5E",              # Vector Rose
                "XGBoost": "#0EA5E9",          # Boost Sky
                "LSTM": "#8B5CF6",             # Memory Violet
                "Ensemble": "#6366F1",         # Ensemble Indigo
                "Hybrid": "#8B5CF6",           # Hybrid Purple
                "Custom": "#6B7280"            # Custom Gray
            },
            "carriers": {
                "electricity": "#3B82F6",      # Blue
                "natural_gas": "#EF4444",      # Red
                "coal": "#1F2937",             # Gray
                "oil": "#92400E",              # Brown
                "biomass": "#16A34A",          # Green
                "solar": "#F59E0B",            # Amber
                "wind": "#06B6D4",             # Cyan
                "hydro": "#0891B2",            # Light Blue
                "nuclear": "#7C3AED",          # Purple
                "geothermal": "#DC2626",       # Red
                "waste": "#6B7280",            # Gray
                "hydrogen": "#8B5CF6",         # Violet
                "battery": "#10B981",          # Emerald
                "pumped_hydro": "#0D9488",     # Teal
                "compressed_air": "#84CC16",   # Lime
                "thermal": "#F97316",          # Orange
                "district_heating": "#EC4899", # Pink
                "district_cooling": "#14B8A6"  # Teal
            },
            "status": {
                "success": "#10B981",          # Emerald
                "warning": "#F59E0B",          # Amber
                "error": "#EF4444",            # Red
                "info": "#3B82F6",             # Blue
                "pending": "#6B7280",          # Gray
                "active": "#8B5CF6",           # Violet
                "inactive": "#9CA3AF",         # Light Gray
                "completed": "#059669",        # Dark Green
                "failed": "#DC2626",           # Dark Red
                "cancelled": "#F97316"         # Orange
            },
            "charts": {
                "primary": "#2563EB",          # Vibrant Blue
                "secondary": "#059669",        # Emerald Green
                "tertiary": "#DC2626",         # Strong Red
                "quaternary": "#7C3AED",       # Rich Purple
                "quinary": "#EA580C",          # Warm Orange
                "senary": "#0891B2",           # Steel Blue
                "septenary": "#BE185D",        # Magenta Pink
                "octonary": "#16A34A",         # Forest Green
                "nonary": "#F59E0B",           # Golden Amber
                "denary": "#8B5CF6",           # Violet
                "background": "#FFFFFF",       # Pure White
                "surface": "#F8FAFC",          # Light Surface
                "grid": "rgba(226, 232, 240, 0.8)",  # Subtle Grid
                "text": "#1E293B",             # Dark Text
                "axis": "#64748B",             # Axis Gray
                "hover": "rgba(241, 245, 249, 0.9)",  # Hover Overlay
                "border": "#E2E8F0",           # Border Gray
                "shadow": "rgba(0, 0, 0, 0.1)" # Subtle Shadow
            },
            "gradients": {
                "primary": ["#3B82F6", "#1E40AF", "#1D4ED8"],        # Blue gradient
                "secondary": ["#10B981", "#047857", "#059669"],       # Green gradient
                "tertiary": ["#EF4444", "#B91C1C", "#DC2626"],        # Red gradient
                "quaternary": ["#8B5CF6", "#6D28D9", "#7C3AED"],      # Purple gradient
                "success": ["#10B981", "#047857", "#059669"],         # Success gradient
                "warning": ["#F59E0B", "#B45309", "#D97706"],         # Warning gradient
                "error": ["#EF4444", "#B91C1C", "#DC2626"],           # Error gradient
                "info": ["#3B82F6", "#1E40AF", "#2563EB"],            # Info gradient
                "chart_area": ["rgba(59, 130, 246, 0.1)", "rgba(59, 130, 246, 0.3)"],  # Chart area gradient
                "chart_line": ["#2563EB", "#1D4ED8"],                 # Chart line gradient
                "heatmap_cool": ["#EFF6FF", "#DBEAFE", "#BFDBFE", "#93C5FD", "#60A5FA", "#3B82F6"],  # Cool heatmap
                "heatmap_warm": ["#FEF3C7", "#FDE68A", "#FCD34D", "#FBBF24", "#F59E0B", "#D97706"]   # Warm heatmap
            },
            "themes": {
                "light": {
                    "background": "#FFFFFF",
                    "surface": "#F8FAFC",
                    "primary": "#2563EB",
                    "secondary": "#64748B",
                    "text": "#1E293B",
                    "border": "#E2E8F0"
                },
                "dark": {
                    "background": "#0F172A",
                    "surface": "#1E293B",
                    "primary": "#3B82F6",
                    "secondary": "#94A3B8",
                    "text": "#F1F5F9",
                    "border": "#334155"
                }
            }
        }
    
    def load_colors(self):
        """Load colors from JSON file or create default"""
        try:
            if os.path.exists(self.config_file):
                with open(self.config_file, 'r') as f:
                    saved_colors = json.load(f)
                # Merge with defaults to ensure all categories exist
                self.colors = self._merge_colors(self.default_colors, saved_colors)
                logger.info(f"Colors loaded from {self.config_file}")
            else:
                self.colors = self.default_colors.copy()
                self.save_colors()
                logger.info("Default colors created and saved")
        except Exception as e:
            logger.error(f"Error loading colors: {e}")
            self.colors = self.default_colors.copy()
    
    def save_colors(self):
        """Save current colors to JSON file"""
        try:
            with open(self.config_file, 'w') as f:
                json.dump(self.colors, f, indent=2)
            logger.info(f"Colors saved to {self.config_file}")
            return True
        except Exception as e:
            logger.error(f"Error saving colors: {e}")
            return False
    
    def _merge_colors(self, defaults: Dict, saved: Dict) -> Dict:
        """Merge saved colors with defaults to ensure completeness"""
        merged = defaults.copy()
        for category, colors in saved.items():
            if category in merged and isinstance(colors, dict):
                merged[category].update(colors)
            else:
                merged[category] = colors
        return merged
    
    def get_color(self, category: str, item: str, default: str = "#6B7280") -> str:
        """Get color for specific category and item"""
        try:
            return self.colors.get(category, {}).get(item, default)
        except Exception:
            return default
    
    def get_color_palette(self, category: str, items: List[str]) -> Dict[str, str]:
        """Get color palette for multiple items in a category"""
        palette = {}
        category_colors = self.colors.get(category, {})
        
        for item in items:
            if item in category_colors:
                palette[item] = category_colors[item]
            else:
                # Generate color if not exists
                palette[item] = self._generate_color_for_item(category, item)
                # Save the generated color
                if category not in self.colors:
                    self.colors[category] = {}
                self.colors[category][item] = palette[item]
        
        return palette
    
    def _generate_color_for_item(self, category: str, item: str) -> str:
        """Generate a consistent, visually appealing color for an item based on its name"""
        import hashlib
        
        # Use hash of item name to generate consistent color
        hash_object = hashlib.md5(f"{category}_{item}".encode())
        hex_dig = hash_object.hexdigest()
        
        # Convert to HSL for better color control
        hue = int(hex_dig[:3], 16) % 360  # 0-359 degrees
        
        # Use predefined saturation and lightness for better visual appeal
        saturation = 65 + (int(hex_dig[3:5], 16) % 25)  # 65-90%
        lightness = 45 + (int(hex_dig[5:7], 16) % 20)   # 45-65%
        
        # Convert HSL to RGB
        def hsl_to_rgb(h, s, l):
            h = h / 360
            s = s / 100
            l = l / 100
            
            def hue_to_rgb(p, q, t):
                if t < 0: t += 1
                if t > 1: t -= 1
                if t < 1/6: return p + (q - p) * 6 * t
                if t < 1/2: return q
                if t < 2/3: return p + (q - p) * (2/3 - t) * 6
                return p
            
            if s == 0:
                r = g = b = l  # achromatic
            else:
                q = l * (1 + s) if l < 0.5 else l + s - l * s
                p = 2 * l - q
                r = hue_to_rgb(p, q, h + 1/3)
                g = hue_to_rgb(p, q, h)
                b = hue_to_rgb(p, q, h - 1/3)
            
            return int(r * 255), int(g * 255), int(b * 255)
        
        r, g, b = hsl_to_rgb(hue, saturation, lightness)
        return f"#{r:02X}{g:02X}{b:02X}"
    
    def set_color(self, category: str, item: str, color: str) -> bool:
        """Set color for specific category and item"""
        try:
            if category not in self.colors:
                self.colors[category] = {}
            self.colors[category][item] = color
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error setting color: {e}")
            return False
    
    def set_colors(self, category: str, color_dict: Dict[str, str]) -> bool:
        """Set multiple colors for a category"""
        try:
            if category not in self.colors:
                self.colors[category] = {}
            self.colors[category].update(color_dict)
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error setting colors: {e}")
            return False
    
    def get_all_colors(self) -> Dict:
        """Get all colors"""
        return self.colors.copy()
    
    def get_category_colors(self, category: str) -> Dict[str, str]:
        """Get all colors for a specific category"""
        return self.colors.get(category, {}).copy()
    
    def reset_to_defaults(self, category: Optional[str] = None) -> bool:
        """Reset colors to defaults"""
        try:
            if category:
                if category in self.default_colors:
                    self.colors[category] = self.default_colors[category].copy()
            else:
                self.colors = self.default_colors.copy()
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error resetting colors: {e}")
            return False
    
    def get_chart_colors(self, count: int, category: str = "charts") -> List[str]:
        """Get an enhanced list of colors for charts with better visual distribution"""
        base_colors = [
            self.get_color(category, "primary"),     # Blue
            self.get_color(category, "secondary"),   # Green
            self.get_color(category, "tertiary"),    # Red
            self.get_color(category, "quaternary"),  # Purple
            self.get_color(category, "quinary"),     # Orange
            self.get_color(category, "senary"),      # Steel Blue
            self.get_color(category, "septenary"),   # Magenta
            self.get_color(category, "octonary"),    # Forest Green
            self.get_color(category, "nonary"),      # Amber
            self.get_color(category, "denary")       # Violet
        ]
        
        # If we need more colors, generate them with better distribution
        if count > len(base_colors):
            additional_colors = self._generate_additional_colors(count - len(base_colors))
            base_colors.extend(additional_colors)
        
        return base_colors[:count]
    
    def _generate_additional_colors(self, count: int) -> List[str]:
        """Generate additional colors with good visual distribution"""
        # Enhanced color palette for additional colors
        extended_palette = [
            "#6366F1",  # Indigo
            "#14B8A6",  # Teal
            "#F43F5E",  # Rose
            "#84CC16",  # Lime
            "#F97316",  # Orange
            "#EC4899",  # Pink
            "#06B6D4",  # Cyan
            "#A855F7",  # Purple
            "#22C55E",  # Green
            "#EF4444",  # Red
            "#8B5CF6",  # Violet
            "#10B981",  # Emerald
            "#F59E0B",  # Amber
            "#3B82F6",  # Blue
            "#6B7280"   # Gray
        ]
        
        colors = []
        for i in range(count):
            if i < len(extended_palette):
                colors.append(extended_palette[i])
            else:
                # Generate color using improved algorithm
                colors.append(self._generate_color_for_item("chart", f"extended_{i}"))
        
        return colors
    
    def get_sector_colors(self, sectors: List[str]) -> Dict[str, str]:
        """Get colors specifically for sectors"""
        return self.get_color_palette("sectors", sectors)
    
    def get_model_colors(self, models: List[str]) -> Dict[str, str]:
        """Get colors specifically for models"""
        return self.get_color_palette("models", models)
    
    def get_carrier_colors(self, carriers: List[str]) -> Dict[str, str]:
        """Get colors specifically for carriers"""
        return self.get_color_palette("carriers", carriers)
    
    def export_colors_for_js(self) -> str:
        """Export colors as JavaScript object"""
        try:
            return f"window.AppColors = {json.dumps(self.colors, indent=2)};"
        except Exception as e:
            logger.error(f"Error exporting colors for JS: {e}")
            return "window.AppColors = {};"
    
    def get_gradient(self, gradient_name: str) -> List[str]:
        """Get gradient colors"""
        return self.colors.get("gradients", {}).get(gradient_name, ["#3B82F6", "#1D4ED8"])
    
    def get_theme_colors(self, theme: str = "light") -> Dict[str, str]:
        """Get theme colors"""
        return self.colors.get("themes", {}).get(theme, self.colors["themes"]["light"])

# Global instance
color_manager = ColorManager()

def init_color_manager(app):
    """Initialize color manager with Flask app"""
    color_manager.init_app(app)
    return color_manager

# Utility functions for direct use
def get_color(category: str, item: str, default: str = "#6B7280") -> str:
    """Direct function to get color"""
    return color_manager.get_color(category, item, default)

def get_sector_colors(sectors: List[str]) -> Dict[str, str]:
    """Direct function to get sector colors"""
    return color_manager.get_sector_colors(sectors)

def get_model_colors(models: List[str]) -> Dict[str, str]:
    """Direct function to get model colors"""
    return color_manager.get_model_colors(models)

def get_chart_colors(count: int) -> List[str]:
    """Direct function to get chart colors"""
    return color_manager.get_chart_colors(count)

File: static/js/color_manager.js
/**
 * JavaScript Color Manager for Energy Platform
 * Frontend color management system that syncs with backend ColorManager
 */

class ColorManager {
    constructor() {
        this.colors = {};
        this.initialized = false;
        this.apiBase = '/api/colors';
        this.loadColors();
    }

    /**
     * Initialize color manager and load colors
     */
    async loadColors() {
        try {
            // First try to get colors from window.AppColors (set by template)
            if (window.AppColors && typeof window.AppColors === 'object') {
                this.colors = window.AppColors;
                this.initialized = true;
                console.log('✓ Colors loaded from window.AppColors');
                return;
            }

            // Fallback to API if window.AppColors not available
            const response = await fetch(`${this.apiBase}/get-all`);
            if (response.ok) {
                const result = await response.json();
                if (result.status === 'success' && result.data) {
                    this.colors = result.data;
                    this.initialized = true;
                    console.log('✓ Colors loaded from API');
                } else {
                    this.loadDefaultColors();
                    console.warn(`⚠ Using default colors - API error: ${result.message || 'Unknown error'}`);
                }
            } else {
                this.loadDefaultColors();
                console.warn('⚠ Using default colors - API not available');
            }
        } catch (error) {
            console.error('Error loading colors:', error);
            this.loadDefaultColors();
        }
    }

    /**
     * Load default colors as fallback
     */
    loadDefaultColors() {
        this.colors = {
            sectors: {
                residential: "#2563EB",
                commercial: "#059669",
                industrial: "#DC2626",
                transportation: "#7C3AED",
                agriculture: "#16A34A",
                public: "#EA580C",
                mining: "#92400E",
                construction: "#0891B2",
                services: "#BE185D",
                other: "#6B7280"
            },
            models: {
                MLR: "#3B82F6",
                SLR: "#10B981",
                WAM: "#F59E0B",
                TimeSeries: "#8B5CF6",
                ARIMA: "#EF4444",
                Linear: "#06B6D4"
            },
            carriers: {
                electricity: "#3B82F6",
                natural_gas: "#EF4444",
                coal: "#1F2937",
                oil: "#92400E",
                biomass: "#16A34A",
                solar: "#F59E0B",
                wind: "#06B6D4",
                hydro: "#0891B2",
                nuclear: "#7C3AED"
            },
            status: {
                success: "#10B981",
                warning: "#F59E0B",
                error: "#EF4444",
                info: "#3B82F6",
                pending: "#6B7280"
            },
            charts: {
                primary: "#2563EB",
                secondary: "#059669",
                tertiary: "#DC2626",
                quaternary: "#7C3AED",
                quinary: "#EA580C"
            }
        };
        this.initialized = true;
    }

    /**
     * Get color for specific category and item
     */
    getColor(category, item, defaultColor = "#6B7280") {
        if (!this.initialized) {
            console.warn('ColorManager not initialized, using default color');
            return defaultColor;
        }

        try {
            const categoryColors = this.colors[category];
            if (categoryColors && categoryColors[item]) {
                return categoryColors[item];
            }

            // Generate color if not exists
            if (categoryColors) {
                const generatedColor = this.generateColorForItem(category, item);
                categoryColors[item] = generatedColor;
                return generatedColor;
            }

            return defaultColor;
        } catch (error) {
            console.error(`Error getting color for ${category}.${item}:`, error);
            return defaultColor;
        }
    }

    /**
     * Get color palette for multiple items in a category
     */
    getColorPalette(category, items) {
        const palette = {};
        items.forEach(item => {
            palette[item] = this.getColor(category, item);
        });
        return palette;
    }

    /**
     * Get colors specifically for sectors
     */
    getSectorColors(sectors) {
        return this.getColorPalette('sectors', sectors);
    }

    /**
     * Get colors specifically for models
     */
    getModelColors(models) {
        return this.getColorPalette('models', models);
    }

    /**
     * Get colors specifically for carriers
     */
    getCarrierColors(carriers) {
        return this.getColorPalette('carriers', carriers);
    }

    /**
     * Get chart colors for specified count
     */
    getChartColors(count, category = 'charts') {
        const baseColors = [
            this.getColor(category, 'primary'),
            this.getColor(category, 'secondary'),
            this.getColor(category, 'tertiary'),
            this.getColor(category, 'quaternary'),
            this.getColor(category, 'quinary')
        ];

        if (count <= baseColors.length) {
            return baseColors.slice(0, count);
        }

        // Generate additional colors if needed
        const additionalColors = [];
        for (let i = baseColors.length; i < count; i++) {
            additionalColors.push(this.generateColorForItem('chart', `color_${i}`));
        }

        return [...baseColors, ...additionalColors];
    }

    /**
     * Generate consistent color for an item based on its name
     */
    generateColorForItem(category, item) {
        // Simple hash function for consistent color generation
        let hash = 0;
        const str = `${category}_${item}`;
        for (let i = 0; i < str.length; i++) {
            hash = ((hash << 5) - hash + str.charCodeAt(i)) & 0xffffffff;
        }

        // Convert to positive number and generate RGB
        hash = Math.abs(hash);
        const r = Math.floor((hash % 256));
        const g = Math.floor(((hash >> 8) % 256));
        const b = Math.floor(((hash >> 16) % 256));

        // Ensure good contrast and visibility
        const brightness = (r * 299 + g * 587 + b * 114) / 1000;
        let adjustedR = r, adjustedG = g, adjustedB = b;

        if (brightness < 100) { // Too dark
            adjustedR = Math.min(255, r + 100);
            adjustedG = Math.min(255, g + 100);
            adjustedB = Math.min(255, b + 100);
        } else if (brightness > 200) { // Too light
            adjustedR = Math.max(0, r - 100);
            adjustedG = Math.max(0, g - 100);
            adjustedB = Math.max(0, b - 100);
        }

        return `#${adjustedR.toString(16).padStart(2, '0')}${adjustedG.toString(16).padStart(2, '0')}${adjustedB.toString(16).padStart(2, '0')}`;
    }

    /**
     * Set color for specific category and item
     */
    async setColor(category, item, color) {
        try {
            if (!this.colors[category]) {
                this.colors[category] = {};
            }
            this.colors[category][item] = color;

            // Try to save to backend
            const response = await fetch(`${this.apiBase}/set`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    category: category,
                    item: item,
                    color: color
                })
            });

            if (!response.ok) {
                console.warn('Failed to save color to backend');
            }

            return true;
        } catch (error) {
            console.error('Error setting color:', error);
            return false;
        }
    }

    /**
     * Set multiple colors for a category
     */
    async setColors(category, colorDict) {
        try {
            if (!this.colors[category]) {
                this.colors[category] = {};
            }
            Object.assign(this.colors[category], colorDict);

            // Try to save to backend
            const response = await fetch(`${this.apiBase}/set-multiple`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    category: category,
                    colors: colorDict
                })
            });

            if (!response.ok) {
                console.warn('Failed to save colors to backend');
            }

            return true;
        } catch (error) {
            console.error('Error setting colors:', error);
            return false;
        }
    }

    /**
     * Add transparency to hex color
     */
    addTransparency(hexColor, alpha) {
        try {
            const hex = hexColor.replace('#', '');
            const r = parseInt(hex.substr(0, 2), 16);
            const g = parseInt(hex.substr(2, 2), 16);
            const b = parseInt(hex.substr(4, 2), 16);
            return `rgba(${r}, ${g}, ${b}, ${alpha})`;
        } catch (error) {
            return `rgba(59, 130, 246, ${alpha})`;
        }
    }

    /**
     * Darken a hex color by a factor
     */
    darkenColor(hexColor, factor) {
        try {
            const hex = hexColor.replace('#', '');
            let r = parseInt(hex.substr(0, 2), 16);
            let g = parseInt(hex.substr(2, 2), 16);
            let b = parseInt(hex.substr(4, 2), 16);

            r = Math.max(0, Math.floor(r * (1 - factor)));
            g = Math.max(0, Math.floor(g * (1 - factor)));
            b = Math.max(0, Math.floor(b * (1 - factor)));

            return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
        } catch (error) {
            return '#1D4ED8';
        }
    }

    /**
     * Lighten a hex color by a factor
     */
    lightenColor(hexColor, factor) {
        try {
            const hex = hexColor.replace('#', '');
            let r = parseInt(hex.substr(0, 2), 16);
            let g = parseInt(hex.substr(2, 2), 16);
            let b = parseInt(hex.substr(4, 2), 16);

            r = Math.min(255, Math.floor(r + (255 - r) * factor));
            g = Math.min(255, Math.floor(g + (255 - g) * factor));
            b = Math.min(255, Math.floor(b + (255 - b) * factor));

            return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
        } catch (error) {
            return '#93C5FD';
        }
    }

    /**
     * Get gradient colors
     */
    getGradient(gradientName) {
        const gradients = this.colors.gradients || {};
        return gradients[gradientName] || ["#3B82F6", "#1D4ED8"];
    }

    /**
     * Create Chart.js dataset with proper colors
     */
    createDataset(label, data, options = {}) {
        const {
            category = 'charts',
            item = 'primary',
            type = 'line',
            fill = false,
            tension = 0.1
        } = options;

        const color = this.getColor(category, item);

        const dataset = {
            label: label,
            data: data,
            borderColor: color,
            backgroundColor: fill ? this.addTransparency(color, 0.3) : this.addTransparency(color, 0.1),
            fill: fill,
            tension: tension,
            borderWidth: 2,
            pointRadius: 3,
            pointHoverRadius: 5
        };

        // Type-specific styling
        if (type === 'bar') {
            dataset.backgroundColor = color;
            dataset.borderWidth = 1;
        } else if (type === 'area') {
            dataset.fill = true;
            dataset.backgroundColor = this.addTransparency(color, 0.3);
        }

        return dataset;
    }

    /**
     * Get all colors
     */
    getAllColors() {
        return { ...this.colors };
    }

    /**
     * Get category colors
     */
    getCategoryColors(category) {
        return { ...(this.colors[category] || {}) };
    }

    /**
     * Wait for initialization
     */
    async waitForInitialization() {
        if (this.initialized) return;

        let attempts = 0;
        const maxAttempts = 50;

        while (!this.initialized && attempts < maxAttempts) {
            await new Promise(resolve => setTimeout(resolve, 100));
            attempts++;
        }

        if (!this.initialized) {
            console.warn('ColorManager initialization timeout, using defaults');
            this.loadDefaultColors();
        }
    }

    /**
     * Reset colors to defaults
     */
    async resetToDefaults(category = null) {
        try {
            const response = await fetch(`${this.apiBase}/reset`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ category: category })
            });

            if (response.ok) {
                await this.loadColors();
                return true;
            } else {
                console.warn('Failed to reset colors via API');
                if (category && this.colors[category]) {
                    // Reset locally
                    this.loadDefaultColors();
                    return true;
                }
            }
        } catch (error) {
            console.error('Error resetting colors:', error);
        }
        return false;
    }

    /**
     * Save all current colors to backend
     */
    async saveAllColors() {
        try {
            const response = await fetch(`${this.apiBase}/save-all`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(this.colors)
            });

            if (!response.ok) {
                const errorData = await response.json().catch(() => ({ message: `Server responded with status ${response.status}` }));
                console.error('Failed to save all colors to backend:', errorData);
                throw new Error(errorData.message || `Failed to save colors. Status: ${response.status}`);
            }

            const result = await response.json();
            if (result.status !== 'success') {
                throw new Error(result.message || 'Backend reported an error on save.');
            }

            return true;
        } catch (error) {
            console.error('Error in saveAllColors:', error);
            throw error; // Re-throw to be caught by the caller
        }
    }
}

// Create global instance
window.ColorManager = ColorManager;
window.colorManager = new window.ColorManager();

// Utility functions for global access
window.getColor = (category, item, defaultColor) => window.colorManager.getColor(category, item, defaultColor);
window.getSectorColors = (sectors) => window.colorManager.getSectorColors(sectors);
window.getModelColors = (models) => window.colorManager.getModelColors(models);
window.getChartColors = (count) => window.colorManager.getChartColors(count);

// Export for modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ColorManager;
}